### PR TITLE
Refactored code to use StatsRecorder with all the frontends

### PR DESCRIPTION
**Pull request by pandurangpatil**
Refactored code to use StatsRecorder with all the frontends

### CODE REVIEW

**Review by khemrajrathore — COMMENTED** [review]
@pandurangpatil Can we think of `StatsRecorder` as a trait instead of a class, which ever pass or class needs to use this functionality can do
`class IdentifierTagger(cpg: Cpg) with StatsRecorder`

What is your thought on this?

**Comment by pandurangpatil** [comment]
> @pandurangpatil Can we think of `StatsRecorder` as a trait instead of a class, which ever pass or class needs to use this functionality can do `class IdentifierTagger(cpg: Cpg) with StatsRecorder`
> 
> What is your thought on this?

No.  We need a single instance to be used across, as it internally holds a Writer thread which will write stats after configured frequency.

File: src/main/scala/ai/privado/entrypoint/Main.scala [hunk 1]
```diff
@@ -25,20 +25,38 @@ package ai.privado.entrypoint
...
 import scala.sys.exit
 
 /** Privado Core main entry point
   */
-object Main {
+object Main extends GeneralMetadataLoggers {
   private val logger = LoggerFactory.getLogger(this.getClass)
 
   def main(args: Array[String]): Unit = {
+    val statsRecorder = StatsRecorder(fullName = true, printToConsole = true, supressSubstages = false)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/entrypoint/Main.scala:40]: change this `printToConsole = false`, this should avoid printing the console logs for joern passes while generating the CPG. 

Test the change locally by running a scan with any sample repo.

File: src/main/scala/ai/privado/languageEngine/csharp/processor/CSharpProcessor.scala [hunk 2]
```diff
@@ -123,11 +114,12 @@ class CSharpProcessor(
...
     val xtocpg = new CSharpSrc2Cpg().createCpg(cpgconfig).map { cpg =>
-      println(
-        s"${TimeMetric.getNewTime()} - Base processing done in \t\t\t\t- ${TimeMetric.setNewTimeToLastAndGetTimeDiff()}"
-      )
+      statsRecorder.endLastStage()
+      statsRecorder.initiateNewStage("Default overlays")
       // Apply default overlays
       X2Cpg.applyDefaultOverlays(cpg)
+      statsRecorder.endLastStage()
+      statsRecorder.setSupressSubstagesFlag(false)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/csharp/processor/CSharpProcessor.scala:122]: In continuation with this comment - https://github.com/Privado-Inc/privado-core/pull/1073/files#r1644326841

Here we are enabling again to print the passes stats (start and end event) to be printed again. As from this point onwards all the passes are of privado.

I don't see these statements for other language processors, can you check and add them?

File: src/main/scala/ai/privado/languageEngine/default/processor/DefaultProcessor.scala [hunk 3]
```diff
@@ -105,11 +102,6 @@ class DefaultProcessor(
 
     val xtocpg = withNewEmptyCpg(cpgOutputPath, cpgconfig: JavaConfig) { (cpg, config) =>
       {
-        println(
-          s"${TimeMetric.getNewTime()} - Base processing done in \t\t\t\t- ${TimeMetric.setNewTimeToLastAndGetTimeDiff()}"
-        )
-        // Apply default overlays
-        X2Cpg.applyDefaultOverlays(cpg)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/default/processor/DefaultProcessor.scala:112]: I think applying default overlays is not required for default processor but cross check it once.

File: src/main/scala/ai/privado/languageEngine/go/processor/GoProcessor.scala [hunk 4]
```diff
@@ -49,21 +62,22 @@ object GoProcessor {
...
-          if (ScanProcessor.config.enableLambdaFlows)
+          if (ScanProcessor.config.enableLambdaFlows) {
             new ExperimentalLambdaDataFlowSupportPass(cpg).createAndApply()
-          logger.info("=====================")
-          println(
-            s"${TimeMetric.getNewTime()} - Run oss data flow is done in \t\t\t- ${TimeMetric.setNewTimeToLastAndGetTimeDiff()}"
-          )
+          }
+          statsRecorder.endLastStage()
+          statsRecorder.setSupressSubstagesFlag(false)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/go/processor/GoProcessor.scala:78]: For go we are enabling it at this point

File: src/main/scala/ai/privado/languageEngine/java/processor/JavaProcessor.scala [hunk 5]
```diff
@@ -159,18 +154,22 @@ class JavaProcessor(
...
-
+      statsRecorder.endLastStage()
+      statsRecorder.initiateNewStage("Preprocessing before overlay passes")
       new LoggerLombokPass(cpg).createAndApply()
-
+      statsRecorder.endLastStage()
+      statsRecorder.initiateNewStage("Default overlays")
       applyDefaultOverlays(cpg)
+      statsRecorder.endLastStage()
+      statsRecorder.setSupressSubstagesFlag(false)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/java/processor/JavaProcessor.scala:172]: For Java frontend its done

File: src/main/scala/ai/privado/languageEngine/java/tagger/sink/api/JavaAPITagger.scala [hunk 6]
```diff
@@ -163,7 +163,7 @@ class JavaAPITagger(cpg: Cpg, ruleCache: RuleCache, privadoInputConfig: PrivadoI
     apiTaggerToUse match {
       case APITaggerVersionJava.V1Tagger =>
         logger.debug("Using brute API Tagger to find API sinks")
-        println(s"${Calendar.getInstance().getTime} - --API TAGGER V1 invoked...")
+        println(s"${Calendar.getInstance().getTime} - -- API TAGGER V1 invoked...")
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/java/tagger/sink/api/JavaAPITagger.scala:166]: Use `statsRecorder.justLog`

File: src/main/scala/ai/privado/languageEngine/java/tagger/sink/api/JavaAPITagger.scala [hunk 7]
```diff
@@ -185,7 +185,7 @@ class JavaAPITagger(cpg: Cpg, ruleCache: RuleCache, privadoInputConfig: PrivadoI
         )
       case APITaggerVersionJava.V2Tagger =>
         logger.debug("Using Enhanced API tagger to find API sinks")
-        println(s"${Calendar.getInstance().getTime} - --API TAGGER V2 invoked...")
+        println(s"${Calendar.getInstance().getTime} - -- API TAGGER V2 invoked...")
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/java/tagger/sink/api/JavaAPITagger.scala:188]: Use `statsRecorder.justLog`

File: src/main/scala/ai/privado/languageEngine/java/tagger/sink/api/JavaAPITagger.scala [hunk 8]
```diff
@@ -197,7 +197,7 @@ class JavaAPITagger(cpg: Cpg, ruleCache: RuleCache, privadoInputConfig: PrivadoI
         )
       case _ =>
         logger.debug("Skipping API Tagger because valid match not found, only applying Feign client")
-        println(s"${Calendar.getInstance().getTime} - --API TAGGER SKIPPED, applying Feign client API...")
+        println(s"${Calendar.getInstance().getTime} - -- API TAGGER SKIPPED, applying Feign client API...")
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/java/tagger/sink/api/JavaAPITagger.scala:200]: Use `statsRecorder.justLog`

File: src/main/scala/ai/privado/languageEngine/javascript/processor/JavascriptProcessor.scala [hunk 9]
```diff
@@ -125,10 +129,7 @@ class JavascriptProcessor(
       .withIgnoredFilesRegex(ruleCache.getExclusionRegex)
 
     val xtocpg = new JsSrc2Cpg().createCpg(cpgConfig).map { cpg =>
-      println(
-        s"${TimeMetric.getNewTime()} - Base processing done in \t\t\t\t- ${TimeMetric.setNewTimeToLastAndGetTimeDiff()}"
-      )
-
+      statsRecorder.endLastStage()
       applyDefaultOverlays(cpg)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/javascript/processor/JavascriptProcessor.scala:133]: Need to start a stage before `applyDefaultOverlays(cpg)` and end after it.

We need to also check for this frontend to enable sub stage stats logging flag.

File: src/main/scala/ai/privado/languageEngine/kotlin/processor/KotlinProcessor.scala [hunk 10]
```diff
@@ -135,11 +122,12 @@ class KotlinProcessor(
...
     val xtocpg = new Kotlin2Cpg().createCpg(cpgconfig).map { cpg =>
-      println(
-        s"${TimeMetric.getNewTime()} - Base processing done in \t\t\t\t- ${TimeMetric.setNewTimeToLastAndGetTimeDiff()}"
-      )
+      statsRecorder.endLastStage()
+      statsRecorder.initiateNewStage("Default overlays")
       // Apply default overlays
       X2Cpg.applyDefaultOverlays(cpg)
+      statsRecorder.endLastStage()
+      statsRecorder.setSupressSubstagesFlag(false)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/kotlin/processor/KotlinProcessor.scala:130]: substage loggging enable for this frontend.

File: src/main/scala/ai/privado/languageEngine/php/processor/PhpProcessor.scala [hunk 11]
```diff
@@ -127,7 +131,7 @@ object PhpProcessor {
       case None    => Paths.get(fixedDir, "bin", "php-parser", "php-parser.php")
     }
 
-    println(s"${TimeMetric.getNewTime()} - Using PHP logger from $parserPath")
+    println(s"${Calendar.getInstance().getTime} - Using PHP logger from $parserPath")
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/php/processor/PhpProcessor.scala:134]: use `statsRecorder.justLog`

File: src/main/scala/ai/privado/languageEngine/php/processor/PhpProcessor.scala [hunk 12]
```diff
@@ -95,11 +99,11 @@ class PhpProcessor(
...
     val xtocpg = new Php2Cpg().createCpg(cpgConfig).map { cpg =>
-      println(
-        s"${TimeMetric.getNewTime()} - Base processing done in \t\t\t\t- ${TimeMetric.setNewTimeToLastAndGetTimeDiff()}"
-      )
-
+      statsRecorder.endLastStage()
+      statsRecorder.initiateNewStage("Default overlays")
       applyDefaultOverlays(cpg)
+      statsRecorder.endLastStage()
+      statsRecorder.setSupressSubstagesFlag(false)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/php/processor/PhpProcessor.scala:106]: for this frontend substage passes for privado is enabled.

File: src/main/scala/ai/privado/languageEngine/python/processor/PythonProcessor.scala [hunk 13]
```diff
@@ -52,38 +65,35 @@ object PythonProcessor {
...
+          statsRecorder.endLastStage()
+          statsRecorder.initiateNewStage("Overriden passes")
           new ImportsPass(cpg).createAndApply()
           new PythonImportResolverPass(cpg).createAndApply()
           new PythonInheritanceNamePass(cpg).createAndApply()
-          println(
-            s"${TimeMetric.getNewTime()} - Run InheritanceFullNamePass done in \t\t\t- ${TimeMetric.setNewTimeToLastAndGetTimeDiff()}"
-          )
-
-          new HTMLParserPass(cpg, sourceRepoLocation, ruleCache, privadoInputConfig = privadoInput)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/python/processor/PythonProcessor.scala:68]: Looks like we removed this pass. We need to add it back. Looks like we made few changes for customer repo debugging, which we didn't revert back.

File: src/main/scala/ai/privado/languageEngine/python/processor/PythonProcessor.scala [hunk 14]
```diff
@@ -52,38 +65,35 @@ object PythonProcessor {
...
           // Apply OSS Dataflow overlay
           new OssDataFlow(new OssDataFlowOptions()).run(new LayerCreatorContext(cpg))
           if (privadoInput.enableLambdaFlows)
             new ExperimentalLambdaDataFlowSupportPass(cpg).createAndApply()
+          statsRecorder.endLastStage()
+          statsRecorder.setSupressSubstagesFlag(false)
+
+          statsRecorder.initiateNewStage("Privado source passes")
+
+          new HTMLParserPass(cpg, sourceRepoLocation, ruleCache, privadoInputConfig = privadoInput)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/python/processor/PythonProcessor.scala:95]: Ohh Ok we moved this pass post all the base passes are processed.

File: src/main/scala/ai/privado/languageEngine/python/tagger/sink/PythonAPITagger.scala [hunk 15]
```diff
@@ -73,7 +73,7 @@ class PythonAPITagger(cpg: Cpg, ruleCache: RuleCache, privadoInput: PrivadoInput
     }
 
     logger.debug("Using Enhanced API tagger to find API sinks")
-    println(s"${Calendar.getInstance().getTime} - --API TAGGER Common HTTP Libraries Used...")
+    println(s"${Calendar.getInstance().getTime} - -- API TAGGER Common HTTP Libraries Used...")
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/python/tagger/sink/PythonAPITagger.scala:76]: use `statsRecorder.justLog`

File: src/main/scala/ai/privado/languageEngine/ruby/processor/RubyProcessor.scala [hunk 16]
```diff
@@ -135,43 +141,26 @@ object RubyProcessor {
...
           new OssDataFlow(options).run(context)
-          if (ScanProcessor.config.enableLambdaFlows)
+          if (ScanProcessor.config.enableLambdaFlows) {
             new ExperimentalLambdaDataFlowSupportPass(cpg).createAndApply()
-          println(
-            s"${TimeMetric.getNewTime()} - Overlay done in \t\t\t- ${TimeMetric.setNewTimeToLastAndGetTimeDiff()}"
-          )
+          }
+          statsRecorder.endLastStage()
+          statsRecorder.setSupressSubstagesFlag(false)
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/ruby/processor/RubyProcessor.scala:161]: For this frontend we have enabled the substage stats logging

File: src/main/scala/ai/privado/languageEngine/ruby/tagger/sink/APITagger.scala [hunk 17]
```diff
@@ -63,7 +63,7 @@ class APITagger(cpg: Cpg, ruleCache: RuleCache, privadoInput: PrivadoInput, appC
     }
 
     logger.debug("Using Enhanced API tagger to find API sinks")
-    println(s"${Calendar.getInstance().getTime} - --API TAGGER Common HTTP Libraries Used...")
+    println(s"${Calendar.getInstance().getTime} - -- API TAGGER Common HTTP Libraries Used...")
```
Comment (pandurangpatil) [inline src/main/scala/ai/privado/languageEngine/ruby/tagger/sink/APITagger.scala:66]: use `statsRecorder.justLog`

File: src/main/scala/ai/privado/languageEngine/default/processor/DefaultProcessor.scala (hunk 3 above)
Comment (karan-batavia) [reply src/main/scala/ai/privado/languageEngine/default/processor/DefaultProcessor.scala:112]: These overlays are applied in code on the `main` branch.

**Review by ankit-privado — APPROVED** [review]
LGTM

**Pull request merged**